sys/auto_init/can/auto_init_can.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2016 OTA keys S.A.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup auto_init
+ * @{
+ * @file
+ * @brief       initializes can device init function
+ *
+ * @author      Toon Stegen <toon.stegen@altran.com>
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ * @author      Aurelien Gonce <aurelien.gonce@altran.com>
+ * @}
+ */
+
+#include <stdio.h>
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#include "can/dll.h"
+#include "can_conf.h"
+#include "can/config.h"
+#include "can/device.h"
+
+#ifdef MODULE_TRX
+#include "can/trx.h"
+#endif
+
+#ifdef MODULE_CAN_ISOTP
+#include "can/isotp.h"
+#endif
+
+#define STACK_SIZE (THREAD_STACKSIZE_DEFAULT + THREAD_EXTRA_STACKSIZE_PRINTF)
+#define PRIORITY (THREAD_PRIORITY_MAIN - CAN_DLL_NUMOF - 2)
+#define NAME_LENGTH 14
+
+#ifdef MODULE_CAN_LINUX
+#define CAN_LINUX_NAME "can_linux"
+static candev_linux_t can_linux[CANDEV_LINUX_NUMOF];
+#endif
+
+#ifdef MODULE_CAN_ISOTP
+static char isotp_stack[STACK_SIZE];
+#define ISOTP_PRIORITY (PRIORITY + CAN_DLL_NUMOF)
+#endif
+
+static char can_thread_names[CAN_DLL_NUMOF][NAME_LENGTH];
+static candev_param_t can_param[CAN_DLL_NUMOF];
+static char _can_device_stacks[CAN_DLL_NUMOF][STACK_SIZE];
+
+void auto_init_candev(void)
+{
+    DEBUG("auto_init_can: init dll\n");
+    can_dll_init();
+
+#ifdef MODULE_CAN_ISOTP
+    DEBUG("auto_init_can: init isotp\n");
+    isotp_init(isotp_stack, STACK_SIZE, ISOTP_PRIORITY, "isotp");
+#endif
+
+    for (int ifnum = 0; ifnum < (int)CAN_DLL_NUMOF; ifnum++) {
+        can_driver_t driver_type = candev_ifnum[ifnum].driver_type;
+        int config_index = candev_ifnum[ifnum].index;
+        candev_t *device = NULL;
+        char *name = can_thread_names[ifnum];
+
+        switch(driver_type) {
+#ifdef MODULE_CAN_LINUX
+        case CAN_DRIVER_LINUX:
+            candev_linux_init(&can_linux[config_index], &can_config_candev_linux[config_index]);
+            device = (candev_t *)&can_linux[config_index];
+            snprintf(name, NAME_LENGTH - 1, CAN_LINUX_NAME"_%d", config_index);
+            break;
+#endif
+        case CAN_DRIVER_STM32:
+            break;
+        case CAN_DRIVER_MCP2515:
+            break;
```

Comment:
There is something I don't like with the way we did this auto init. The main reason why we put eveything into the same function was to be able to control the order of the interfaces. 
But then we have this big `can_iface_t candev_ifnum[]` array in the can_conf.h file which is included from a lot of places, just to have the CAN_DLL_NUMOF variable defined.

I will think about a better way to initialize everything.